### PR TITLE
Login

### PR DESCRIPTION
**Pull request by JrSchoolAccount**
# Frontend

## Login Page /login

- Error handling and validation
- Login logic

## SignUp Page /registrera

- Error handling and validation
- Registration logic

## Log Out Button
- Simple button to call api/logout

## Profile page
- Simple profile page with logout button

---

# Backend

- Session (Cookies and JWT)
- Sign Up API /api/signup
- Login API /api/login
- Logout API /api/logout
- Middleware to protect /profile

---

# MongoDB

## Users Collection
```json
{
  "_id": {"$oid": "66349ec24d0e094813a01d94"},
  "name": "John Doe",
  "email": "john@example.com",
  "password": "$2b$10$PV9m4WELfhvFdceIONrGbeIeDoGqJr4Gwha1dfk059zao7eH5YZ9e",
  "__v": 0
}

```
---

# Cypress E2E Test
- Register new user test
- Login user test
---

# Screenshots

### Sign up
Screenshot 2024-05-16 at 08 39 18

---

### Login
Screenshot 2024-05-16 at 08 38 17

---

### CODE REVIEW

File: app/layout.tsx [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import { useState } from 'react';
```
Comment (NelsonTheBrave) [inline app/layout.tsx:1]: Why is useState imported here? It doesn't seem to be used in this file?

File: app/ui/signup.tsx [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        </Avatar>
+        <Typography component="h1" variant="h5">
+          Ny medlem
+        </Typography>
+        <Box component="form" onSubmit={handleSubmit} sx={{ mt: 3 }}>
+          <Grid container spacing={2}>
+            <Grid item xs={12}>
+              <TextField
+                autoComplete="given-name"
+                name="name"
```
Comment (NelsonTheBrave) [inline app/ui/signup.tsx:70]: There is something a bit weird with how the name-field works. When loading the signup page it is automatically highlighted, but the placeholder text doesn't show, which means you don't understand you're supposed to write a firstname and lastname unless you click somewhere else first, or write and then delete the text. I guess it's the autofocus property that does this, don't know if there is a way to have autofocus and also show placeholder at the same time

File: cypress/e2e/loginUser.cy.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+    cy.get('input[name="email"]').should('be.visible');
+
+    cy.get('input[name="email"]').type('jon.doe@gmail.com');
+    cy.get('input[name="password"]').type('asdfgh');
+
+    cy.get('button[type="submit"]').click();
+    cy.wait(2000);
+
+    cy.url().should('include', '/profil');
```
Comment (NelsonTheBrave) [inline cypress/e2e/loginUser.cy.ts]: This doesn't work for me... I mean the actual functionality to redirect to /profil doesn't happen when I log in :O I go back to '/'

**Review by NelsonTheBrave — CHANGES_REQUESTED** [review]
Really nice work! Good bunch of code you have in this PR :) 
I can't really comment on the quality of the login functionality, encryption etc since I haven't worked with it myself, but it seems like you have used both hashing and salt for the password checking process which is nice.

I have noted a few issues that I found. The main one I think is that after logging in I neither get redirected to the profile page, nor does the navBar change to represent that the session has started. It seems like that is working for you locally, as I've understood, so I'm not sure what would be the reason.

Another thing that would be nice, but not a must, is if there was some kind of message when you register a new user, and not just redirecting to the startpage. I felt the user experience was a bit uncertain whether something actually happened or not.

I would also put just a little bit of margin between the two buttons in the navbar (logga in and registrera) to make it look nicer :)

Also I realised we should talk in the group about where we put or models, in separate files or in one collected file like we do with the types in definitions.ts. But that is more a general thing than relating to your branch.

Great job!

**Comment by FelixHultman** [comment]
I've look at the code but im not really sure i can give you any good feedback since i dont really understand the code that good, but as Erik said i see you are using alot of the elements we looked at in class, so thats cool. 

Im having kinda the same problem as Erik when im trying out the branch, so i guess it will be easiest if we talk about it the whole group maybe? but it is impressive work, looking forward to you presentation :)

File: app/ui/navbar.tsx [hunk 4]
```diff
@@ -20,31 +18,39 @@ import { Stack } from '@mui/material';
 import { Autocomplete } from '@mui/material';
 import { useState } from 'react';
 import { Movie } from '../lib/definitions';
+import { Login, Logout } from '@mui/icons-material';
+import { Tooltip } from '@mui/material';
+import { usePathname, useRouter } from 'next/navigation';
 
 const links = [
   { name: 'Om oss', href: '/om-oss' },
   { name: 'Aktuella visningar', href: '/aktuella-vinsningar' },
```
Comment (FelixHultman) [inline app/ui/navbar.tsx:27]: this should

Comment (FelixHultman) [inline app/ui/navbar.tsx]: this should be aktuella-visningar! :)

File: app/layout.tsx (hunk 1 above)
Comment (JrSchoolAccount) [reply app/layout.tsx:1]: i removed it.

File: cypress/e2e/loginUser.cy.ts (hunk 3 above)
Comment (JrSchoolAccount) [reply cypress/e2e/loginUser.cy.ts]: Changed :)

**Review by JrSchoolAccount — COMMENTED** [review]
Removed navbar login/logout and moved to a profile page solution instead


https://github.com/JrSchoolAccount/Kino-site-nextjs/assets/144328779/21d8430b-e03e-4ccb-bff7-00d6e5332c0d

**Review by simonpizevski — CHANGES_REQUESTED** [review]
I think that it works very good and the new changes to the profile page works good too!. Great job. 

Im not the best at encryption but good job with the hashed passwords and user creation. 

I think that instead of "mina bokningar" we could name it to "min profil" or something like that.

**Review by FelixHultman — APPROVED** [review]

**Comment by JrSchoolAccount** [comment]
I think i have updated the pr and addressed all the issues found :)

Ready for a final review!

**Comment by FelixHultman** [comment]
> I think i have updated the pr and addressed all the issues found :)
> 
> Ready for a final review!

i cant seem to login, (failed to verify session). 
another small thing is that you cant use å, ä, ö when register a new member( not that important just a note)

**Comment by JrSchoolAccount** [comment]
> > I think i have updated the pr and addressed all the issues found :)
> > Ready for a final review!
> 
> i cant seem to login, (failed to verify session). another small thing is that you cant use å, ä, ö when register a new member( not that important just a note)

@FelixHultman 
I fixed the "åäö" issue :) 
Dont have a login error on my end. do you mind trying again

**Comment by FelixHultman** [comment]
> > > I think i have updated the pr and addressed all the issues found :)
> > > Ready for a final review!
> > 
> > 
> > i cant seem to login, (failed to verify session). another small thing is that you cant use å, ä, ö when register a new member( not that important just a note)
> 
> @FelixHultman I fixed the "åäö" issue :) Dont have a login error on my end. do you mind trying again

i just get failed to verify session when trying to log in, but it might be something else im doing wrong. Lets discuss it at the meeting.

**Review by NelsonTheBrave — APPROVED** [review]
Now it works to log in and access the profile page! Just check spelling of "Mina profil" in the navbar ;)

**Review by simonpizevski — APPROVED** [review]
Works great :)

**Pull request merged**